pkg/sources/s3/s3.go

```diff
@@ -396,6 +412,65 @@ func (s *Source) pageChunker(ctx context.Context, client *s3.S3, chunksChan chan
 	_ = s.jobPool.Wait()
 }
 
+func (s *Source) validateBucketAccess(ctx context.Context, client *s3.S3, roleArn string, buckets []string) []error {
+	shouldHaveAccessToAllBuckets := roleArn == ""
+	wasAbleToListAnyBucket := false
+	var errors []error
+
+	for _, bucket := range buckets {
+		if common.IsDone(ctx) {
+			return nil
+		}
```

Comment:
**Suggestion:** Returning an error here makes more sense in the context of validation imo.

```go
return append(errors, ctx.Err())
```